src/dft/src/ScanReplace.cpp

```diff
@@ -0,0 +1,426 @@
+    sta::LibertyPort* non_scan_cell_port = non_scan_cell_ports_iter.next();
+    sta::LibertyPort* scan_equiv
+        = FindEquivalentPortInScanCell(non_scan_cell_port, scan_cell);
+
+    diff += DifferencePerformancePorts(non_scan_cell_port, scan_equiv);
+  }
+
+  return diff;
+}
+
+// We select the scan_cell that is most similar to the non_scan_cell in
+// performance
+std::unique_ptr<ScanCandidate> SelectBestScanCell(
+    const sta::LibertyCell* non_scan_cell,
+    std::vector<std::unique_ptr<ScanCandidate>>& scan_candidates)
+{
+  std::sort(scan_candidates.begin(),
+            scan_candidates.end(),
+            [&non_scan_cell](const auto& lhs, const auto& rhs) {
+              // We want to keep the difference as close as possible to the
+              // non_scan_cell
+              const double difference_lhs = DifferencePerformanceCells(
+                  non_scan_cell, lhs->getScanCell());
+              const double difference_rhs = DifferencePerformanceCells(
+                  non_scan_cell, rhs->getScanCell());
+              return difference_lhs < difference_rhs;
+            });
+
+  return std::move(scan_candidates.at(0));
+}
+
+}  // namespace
+
+ScanCandidate::ScanCandidate(
+    sta::LibertyCell* scan_cell,
+    std::unordered_map<std::string, std::string> port_mapping)
+    : scan_cell_(scan_cell), port_mapping_(port_mapping)
+{
+}
+
+sta::LibertyCell* ScanCandidate::getScanCell() const
+{
+  return scan_cell_;
+}
+
+const std::unordered_map<std::string, std::string>&
+ScanCandidate::getPortMapping() const
+{
+  return port_mapping_;
+}
+
+void ScanCandidate::debugPrintPortMapping() const
+{
+  std::cout << "Port mapping for cell: " << scan_cell_->name() << std::endl;
+  for (const auto& [from_port, to_port] : port_mapping_) {
+    std::cout << "    " << from_port << " -> " << to_port << std::endl;
+  }
+}
+
+void ScanReplace::collectScanCellAvailable()
+{
+  const sta::dbNetwork* db_network = sta_->getDbNetwork();
+
+  std::vector<sta::LibertyCell*> non_scan_cells;
+
+  // Let's collect scan lib cells and non-scan lib cells availables
+  for (odb::dbLib* lib : db_->getLibs()) {
+    for (odb::dbMaster* master : lib->getMasters()) {
+      // We only care about sequential cells in DFT
+      if (!master->isSequential()) {
+        continue;
+      }
+
+      sta::Cell* master_cell = db_network->dbToSta(master);
+      sta::LibertyCell* liberty_cell = db_network->libertyCell(master_cell);
+
+      if (!liberty_cell) {
+        // Without info about the cell we can't do any replacement
+        continue;
+      }
+
+      // If the lib cell is marked as dontUse, then skip it
+      if (liberty_cell->dontUse()) {
+        continue;
+      }
+
+      // We skip clock gating lib cells
+      if (liberty_cell->isClockGate()) {
+        continue;
+      }
+
+      if (IsScanCell(liberty_cell)) {
+        available_scan_lib_cells_.insert(liberty_cell);
+      } else if (CanBeScanReplace(master, liberty_cell)) {
```

Comment:
Why is this necessary?  The cell has to be sequential to reach here and CanBeScanReplace only excludes buff/inv which are not sequential.